### ISSUE TITLE
Use Observables instead of callbacks?

### ISSUE THREAD

**Issue by zolkis** [w3c/wot-scripting-api#5]
As described [here](https://github.com/w3c/wot/blob/master/scripting/open-issues.md#asynchronous-programming-guidelines) and agreed on the face to face meeting,

> Promises should be used for asynchronous operations with one-shot result (either success or fail), and callbacks (later Observables) should be used for asynchronous operations with multiple-shot results (e.g. open ended protocol requests such as discovery).

This issue is for tracking and discussing the WoT use case for [Observables](https://github.com/tc39/proposal-observable).

**Comment by zolkis** [comment]
Use case: return an `Observable` object for open ended [cancelable] asynchronous operations that provides controls over that operation:
- cancel the operation
- attach a callback for handling streamed or partial results
- attach a callback for handling the end of the operation
- attach an error handler callback.

For example:
```javascript
let subscription = asyncOperation(parameters).subscribe({
    next(val) { console.log("Received next result: " + val); },
    error(err) { console.log("Received an error: " + err); },
    complete() { console.log("Operation complete"); }
});
// ...
subscription.cancel();
```
Alternatively, we could say
```javascript
let subscription = asyncOperation(parameters)
    .on("next", val => { console.log("Received next result: " + val); })
    .on("error", function(err) { console.log("Received an error: " + err); })
    .on("complete", () => { console.log("Operation complete"); })
;
```
Or, 
```javascript
let subscription = asyncOperation(parameters)
    .next(val => { console.log("Received next result: " + val); })
    .complete( () => { console.log("Operation complete"); })
    .catch(err => { console.log("Received an error: " + err); });
;
```

**Comment by zolkis** [comment]
This would be used mainly with discovery operations.

**Comment by zolkis** [comment]
We could do a polyfill for the above options, or continue to use callbacks for discovery, but without the ability to cancel discovery, or to know when discovery has ended (when these are supported by the underlying protocols).

**Comment by h0ru5** [comment]
I am +1 on observables for discovery (and maybe also subscription).

I would stick to the current ECMA discussion. I don't know if there is a polyfill already defined, it seems Angular2 used RxJS, which could be an option.

**Comment by h0ru5** [comment]
Conclusion on call 27-03-2017 is to go for a subset of current ECMA discussion, risking a breakage but  betting on promising candidate.
we check the proposal and decide.

**Comment by h0ru5** [comment]
Link to proposal: https://github.com/tc39/proposal-observable

**Comment by zolkis** [comment]
Leaving this open for further feedback.

**Comment by mmccool** [comment]
Discussion in the meeting on Aug 21 did not reveal any major issue but decided to leave this open for now.

**Comment by danielpeintner** [comment]
I think with the latest updates we can close this issue, correct?

**Comment by zolkis** [comment]
Addressed by #86.
